pkgs/native_assets_cli/lib/src/api/build_config.dart

```diff
@@ -32,168 +45,155 @@ abstract class BuildConfig {
+  /// We're trying to avoid breaking changes. However, in the case that we have
+  /// to, the major version mismatch between the Dart or Flutter SDK and a
+  /// `build.dart` script will lead to a nice error message.
+  static Version get latestVersion => BuildConfigImpl.latestVersion;
+
+  /// Constructs a config by parsing CLI arguments and loading the config file.
+  ///
+  /// The [args] must be commandline arguments.
+  ///
+  /// If provided, [environment] must be a map containing environment variables.
+  /// If not provided, [environment] defaults to [Platform.environment].
+  ///
+  /// If provided, [workingDirectory] is used to resolves paths inside
+  /// [environment].
+  /// If not provided, [workingDirectory] defaults to [Directory.current].
+  ///
+  /// This async constructor is intended to be used directly in CLI files.
+  static Future<BuildConfig> fromArgs(
+    List<String> args, {
+    Map<String, String>? environment,
+    Uri? workingDirectory,
+  }) =>
+      BuildConfigImpl.fromArgs(
+        args,
+        environment: environment,
+        workingDirectory: workingDirectory,
+      );
 
   factory BuildConfig({
     required Uri outDir,
     required String packageName,
     required Uri packageRoot,
     required BuildMode buildMode,
     required Architecture targetArchitecture,
-    required OS targetOs,
+    required OS targetOS,
     IOSSdk? targetIOSSdk,
     int? targetAndroidNdkApi,
     CCompilerConfig? cCompiler,
     required LinkModePreference linkModePreference,
-    Map<String, Metadata>? dependencyMetadata,
+    Map<String, Map<String, Object>>? dependencyMetadata,
+    Iterable<String>? supportedAssetTypes,
   }) =>
-      model.BuildConfig(
+      BuildConfigImpl(
         outDir: outDir,
         packageName: packageName,
         packageRoot: packageRoot,
-        buildMode: buildMode as model.BuildMode,
-        targetArchitecture: targetArchitecture as model.Architecture,
-        targetOs: targetOs as model.OS,
-        targetIOSSdk: targetIOSSdk as model.IOSSdk?,
+        buildMode: buildMode as BuildModeImpl,
+        targetArchitecture: targetArchitecture as ArchitectureImpl,
+        targetOS: targetOS as OSImpl,
+        targetIOSSdk: targetIOSSdk as IOSSdkImpl?,
         targetAndroidNdkApi: targetAndroidNdkApi,
-        cCompiler: cCompiler as model.CCompilerConfig?,
-        linkModePreference: linkModePreference as model.LinkModePreference,
-        dependencyMetadata: dependencyMetadata?.cast(),
+        cCompiler: cCompiler as CCompilerConfigImpl?,
+        linkModePreference: linkModePreference as LinkModePreferenceImpl,
+        dependencyMetadata: dependencyMetadata != null
+            ? {
+                for (final entry in dependencyMetadata.entries)
+                  entry.key: Metadata(entry.value.cast())
+              }
+            : {},
+        supportedAssetTypes: supportedAssetTypes,
       );
 
   factory BuildConfig.dryRun({
```

Comment:
`build.dart` writers should never use these constructors in their scripts.

These constructors are for `build.dart` helper packages authors, they need to be able to unit test their helpers. (As is written in the doc comment.)

> The constructor takes that json data structure. Nothing more

We have that, it's `BuildConfigImpl.fromConfig`, but that's not very nice for helper-package authors to use.

> Also: Imagine, we add later on a config.java entry

If we add it as required it would be breaking, it should probably not be required just like `CCompilerConfig? cCompiler,` is not in the other constructor.
Also, if we just have a JSON thing it and would be required, it would fail while parsing JSON, but now it would be swept under the rug because it's a failure at runtime.
If we add only extension types we do no validation, which basically means nothing is required. That's just identical to adding optional parameters later.

> We should optimize for build.dart authors, if we need something on top for dart build we should have that live somewhere else.

We also have helper package authors that need to test their package. The `native_toolchain_c` package is one of those.

Maybe we should split the API up in three APIs? (1) `build.dart` writers, (2) build.dart-helper-packages writers, (3) backends.

Reply:
I have made the `fromArguments` constructor the default unnamed one now.